### ISSUE TITLE
Telegrams lost

### ISSUE THREAD

**Issue by element2202** [Supergiovane/node-red-contrib-knx-ultimate#145]
Hello & thank you for developing this great node!

**Describe the bug**
I suspect this isn't a bug but perhaps something unique with my setup, or bad config! I had hoped to message you on KNX-user-forum but my German is so bad that I couldn't get past the sign up screen, and Google translate was no help ;) Any tips or advice you can offer would be greatly appreciated.

I am seeing occasional telegrams get lost. I originally noticed this on graphs where I'm logging temperatures from the bus, so I used the Watchdog node to investigate further.

![image](https://user-images.githubusercontent.com/67553956/134869411-cefa846f-f3e4-4a27-9b35-97055d7e46d4.png)

The Watchdog node is confirming that connectivity is lost every day or two, and for a seemingly random period/frequency. The last couple of days look like this:

> time=2021-09-25T18:17:49.134Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-25T18:43:09.163Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-25T19:12:49.211Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-25T19:30:09.231Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-25T20:57:49.350Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-25T21:23:09.379Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-25T22:25:49.480Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-25T22:32:29.488Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-27T01:57:51.554Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-27T02:00:31.556Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-27T02:05:51.562Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-27T02:15:11.570Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-27T04:17:51.681Z check=Eth+KNX description=Watchdog elapsed with no response.
> [....]
> time=2021-09-27T04:20:31.682Z check=Eth+KNX description=Watchdog elapsed with no response.
> 
> time=2021-09-27T06:02:01.793Z check=Eth+KNX description=Watchdog elapsed with no response.
> [...]
> time=2021-09-27T07:45:11.930Z check=Eth+KNX description=Watchdog elapsed with no response.
> 

I have been logging the output since February this year and cannot see any particular pattern.

I have a retry interval of 10 seconds and allow 3 attempts before giving an error so these are not minor blips.

**Setup**
Node-red is running alone on a Raspberry Pi
Knx-Ultimate Version: 1.1.72
IP Gateway: Zennio ZSY-IPI-PL

**Additional info**
The bus also has a Loxone Miniserver and Control 4. Loxone has its own bus interface and provides an IP Gateway for Control 4 (i.e. Knx-Ultimate is the only device using the Zennio gateway (apart from ETS when in use)).

No other devices are seeing instability on the bus.

Please do let me know if you have any ideas or suggestions and thanks again 👍

**Comment by Supergiovane** [comment]
Hello
you're very old! Please update to KNX-Ultimate 1.2.46 and let me know.

**Comment by element2202** [comment]
Ah, yes :flushed: Have updated it and Node-red.

Thanks -- will let you know!

**Comment by element2202** [comment]
Hi,

Everything is now on the latest version 👍 

I reduced the retry count to 1 to try and see the problem sooner and a few hours later we had a blip. The nodes reported disconnected, then connFailCb and then finally the BusError (I assume this came last because of the retry). I have attached the output from the Watchdog node here: [KNXWatchdog-1retry.txt](https://github.com/Supergiovane/node-red-contrib-knx-ultimate/files/7246439/KNXWatchdog-1retry.txt).

I also had Group Monitor running on ETS at the same time and you can see the gap where it became disconnected.

![image](https://user-images.githubusercontent.com/67553956/135142437-1589cde7-4b22-4d60-b719-99de2399d838.png)

Admittedly this was short so I've left everything running in the hope that I get a longer outage. I guess all that is going to show will be repeated BusErrors from the Watchdog and a larger gap on ETS.

I'm wondering if you've ever seen an issue like this before and could point me in the right direction for further testing.

Many thanks :)

**Comment by Supergiovane** [comment]
Hi
You should set it more relaxed. 1 Sec is a lot. 
It can be that a glitch in the lan causes a failed response or it can be that the device you query for response, does limit the amount of telegrams sent in a timeframe. 
You should also investigate your lan, maybe by directly connecting the Knx Interface to nodered. (Using a dumb little switch).  Sometime UDP storming protection kicks in on your switch (do you maybe use Unifi gears?)

**Comment by element2202** [comment]
Hi, 

> You should set it more relaxed. 1 Sec is a lot.

You mean the retrys? OK, I've increased that to 3.

> It can be that a glitch in the lan causes a failed response or it can be that the device you
> query for response, does limit the amount of telegrams sent in a timeframe.

I guess the dimmer which the Watchdog is querying could have a limit but I wouldn't expect all of the other devices to become unresponsive also?

Great idea isolating the Pi and KNX IP Interface onto their own dumb switch. I'll try that tomorrow.

Yes..... entire network is Unifi......!

Thanks again :)

**Comment by Supergiovane** [comment]
Hmm… you have to disable all the fancy stuffs on unifi (optimize blah blah blah ans so on) but firt, try with a dumb switch

**Comment by Supergiovane** [comment]
PS: please also disable the WIFI in the rPI or manually select the ethernet interface on the KNX configuration node.

**Comment by element2202** [comment]
Hi,

I've moved the Pi and the IP Gateway to their own dumb (non-unifi) switch. It's still uplinked to the main network but they are the only devices on the switch. Unfortunately the issue remained.

- I have disabled the firewall on the Pi (although it was fully open on the /24 for our network anyway)
- I have run a ping test from another device to the Pi and waited for a watchdog error but there was no packet loss
- Load averages on the Pi are low

Thank you for the suggestion of selecting eth0 on the KNX config node. There isn't any wifi setup but it was on auto so I've changed that. I am now waiting to see if the issue reoccurs.

Next I'll try:

- Doing a ping test to the IP Gateway and checking there's no packet loss when the Watchdog fails
- Removing the uplink to the rest of the network (or even using a crossover cable to the Pi?!) to fully isolate the setup

Quick question..... I noticed the below on the Watchdog node which appears at every check interval (10 secs):

![image](https://user-images.githubusercontent.com/67553956/135480921-e8238f42-7dd6-4601-a18c-307ad4b7720a.png)

I get this *every time* and I'd assumed it meant that it was running the check, rather than the check actually failing, however could you confirm?

Thanks for your continued support :)

**Comment by element2202** [comment]
Hmmm! It goes from the failed beat almost instantly back to full check level though. There is no delay:

(1/1/13) Full check level Eth+KNX - KNX BUS OK
10 seconds
Check level Eth+KNX, failed beat telegram 1 of 3
0.2 seconds

The interface which we are using here is also used by ETS and it's fine..... In Group Monitor I can see the status request coming from KNX Ultimate to 1/1/13 every 10 seconds. The response is immediate and every time. I can also manually read the group address repeatedly from ETS with no loss.

I have setup a manual group address read and I can inject a read 50x over, and each time the response is received immediately and seen on Group Monitor.

![image](https://user-images.githubusercontent.com/67553956/135492189-225018a5-6489-48d3-96b8-28506eae08d2.png)

It doesn't make sense that the Watchdog node is genuinely seeing loss every 10 seconds. The problem happens every 2-6 hours.... there is no immediate problem as far as I can see.

I have another IP Interface but not a router. It's currently in use by something else but I can switch them over.

I've set the debug level to debug but there are no debug messages for the flow. Am I looking in the right place?

Cheers!

**Comment by Supergiovane** [comment]
Hi
I also have the watchdog node and it works flawlessy. 
I’m running out of ideas. 
As a last resort, you can delete the watchdog and create a new one. Don’t copy/paste the old one, just add a new fresh watchdog.

**Comment by element2202** [comment]
![image](https://user-images.githubusercontent.com/67553956/135497466-1bb9bcea-72fd-463f-8c80-cb9745c3d195.png)

Fresh Watchdog and different group address! 

I genuinely don't think that it's missing a beat -- I can see it sending the read request on the bus, it's fine and gets an immediate response. Its weird that it's every single time... I actually thought that was the expected message when it was sending the telegram.

This isn't the real problem though.... the issue I'm seeing is every few hours the retrys are exceeded and full error is thrown. The nodes all report red and connFailCb/Disconnected is passed through the watchdog. At this point there is a genuine disconnection of some sort and data is lost.

Clearly the two issues must be related although they don't exhibit the same conditions.

Thanks -- I'll keep playing :)

**Comment by Supergiovane** [comment]
Do so:
Put a function node behind the error pin of the watchdog. 
Link the function output to the input of the watchdog. 
In the function node, return a **setconfig** json object. Please look at the help wiki of the watchdog and see how to create the json. 
In this case, every time the watchdog get a connection fail error, the function will force the gateway reconnection. 
We’ll see what happens ..

**Comment by element2202** [comment]
Sorry, I'm not sure what you mean by "error pin"... do you mean match BUSError or NodeError on the msg.type?

I've got the json format for msg.setGatewayConfig so I should be able to follow the rest of the instructions.

Is this in attempt to resolve the weird missed beat issue or the full outage issue? As I mentioned earlier, I can continually read a group address whilst the missed beat comes up without any problems. I don't fully understand what missed beat means but I don't think it's missing anything!!!

**Comment by Supergiovane** [comment]
Hi
Yes, sorry, the intent is to force reconnection after bus error. 
It’s not normal that after 2-6 hours, the while connection goes down.

**Comment by element2202** [comment]
Hi -- thanks, I'll get the auto/force reconnection sorted over the weekend.

I'm still keen to get to the root cause of the issue though so I have installed node-red on a windows laptop to test (same IP gateway as the Pi and ETS). I am waiting for another "full error" to happen, however I immediately noticed the following:

![image](https://user-images.githubusercontent.com/67553956/135642832-9b289170-e6e4-4139-839a-aaa7ce2d71b6.png)

Can I just confirm that this is **not** the expected behaviour when the Watchdog node does its first check? Similar to the Pi, everything is fine and I can see the telegram on the bus but I am seeing this yellow colour message. Every time.

Thanks :)

**Comment by Supergiovane** [comment]
Hi
The first one can be a simple wrong message, because i’ve tested mine and it behaves same way. 
The yellow status every 10 secs is not normal instead. Mine works solid until i pull the eth cable, switch off the knx router or disconnect the red/black tp cable.

**Comment by element2202** [comment]
I wanted to show you the yellow status every 10 secs so I've made a quick video:

https://www.dropbox.com/s/ueco0hvi7vocrp7/Node-RED%20_%20192.168.6.49%20%E2%80%94%20Mozilla%20Firefox%202021-10-02%2014-20-02.mp4?dl=0

I also disabled all of the other flows so only the Watchdog is running. Here is the node-red log.

```
2 Oct 14:42:33 - [info] Starting flows
[info] 10/2/2021, 2:42:33 PM KNXUltimate-config: payload cache set to /home/pi/.node-red/knxultimatestorage/knxpersistvalues
[info] 10/2/2021, 2:42:33 PM KNXUltimate-config: payload cache set to /home/pi/.node-red/knxultimatestorage/knxpersistvalues
[info] 10/2/2021, 2:42:33 PM KNXUltimate-config: Bind KNX Bus to interface : eth0 (Interface's name selected from dropdown list). Node Loxone
[info] 10/2/2021, 2:42:33 PM knxUltimate-config: Will connect in 7 secs. with Loxone
2 Oct 14:42:33 - [info] Started flows
[info] 10/2/2021, 2:42:40 PM knxUltimate-config: perform websocket connection on Loxone
[debug] 10/2/2021, 2:42:40 PM knxUltimate-config: Connected.
[info] 10/2/2021, 2:42:43 PM KNXUltimate-config: Loaded saved GA values 322
[info] 10/2/2021, 2:42:43 PM KNXUltimate-config: Do DoInitialReadFromKNXBusOrFile
```

I am seeing the yellow status using two different IP Gateways and two different node-red installs.

I've left the Watchdog connected to the second IP Gateway to see whether it has any of the full disconnection issues (retry exceeded/red/connFailCb/Disconnected etc. ~6hrs). I'll report back tomorrow.

Thanks again :)

**Comment by Supergiovane** [comment]
Hi Ben
I'm very sorry. I've made you lost your time.
My watchdog does exactly what yours does. Mine was set to 120 seconds, so i never saw the yellow status. Yet is set to 10 secs and it does what yours does.
It's a false warning. I'll fix it with the next version.
Sorry for that!!!!

Please let me know if (with the new IP gateway) you lost connection again with connfailCB.

**Comment by element2202** [comment]
No problem -- I'm just pleased we ruled out that problem :)

Will get back to you on the main issue ASAP!

**Comment by Supergiovane** [comment]
Hi Element. 
any news?

**Comment by element2202** [comment]
Hi Max,

Sorry for the delay in getting back to you but I wanted a week to leave the test running.

Tl;dr there is something wrong with the Zennio IP gateway :)

I changed KNX-Ultimate to use a second gateway. The second gateway only supports a single connection and was used exclusively for Control4, so I swapped Knx-Ultimate and Control4 over.

Within a day Control4 (now on Zennio) reported that it had been disconnected/reconnected to the bus and this pattern continued -- reconnections were at a lower frequency than KNX-Ultimate but I suspect that Control4 is simply more tolerant to packet loss.

KNX-Ultimate (now on second gateway) remained connected throughout, and is still connected right now! Initially we had issues with group address responses not getting back to the nodes (Watchdog remained connected) but we could see the responses on the bus with ETS. To resolve this I increased the telegram delays quite severely which fixed the problem. Speed is not a factor and the amount of concurrent node activity was obviously too much for the second gateway but it is stable :)

With hindsight, I should have swapped the gateways at the very beginning, but with ETS being so solid on the Zennio, it never occurred to me that it could have a problem! I've flashed new firmware onto the Zennio too but that hasn't helped and it's EOL now so probably time to invest in a replacement.

Thank you for working through this with me and thank you for your continued development of KNX-Ultimate 👍
